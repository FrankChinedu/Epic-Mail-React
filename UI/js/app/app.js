/* eslint-disable no-unused-vars */
/* eslint-disable no-undef */
const {
  accessToken, userFirstname, userLastname, userEmail,
} = localStorage;

const token = window.localStorage.getItem('accessToken');

const unSetToken = () => {
  window.localStorage.removeItem('accessToken');
  window.localStorage.removeItem('userEmail');
  window.localStorage.removeItem('userFirstname');
  window.localStorage.removeItem('userLastname');
};

const isAuthenticated = () => {
  if (accessToken === undefined
     && userFirstname === undefined
     && userLastname === undefined && userEmail === undefined) {
    window.location.href = './signin.html';
  }
};
isAuthenticated();


const logout = () => {
  unSetToken();
  isAuthenticated();
  window.location.href = './signin.html';
};

const formatDate = (time) => {
  let t = new Date(time);
  t = t.toLocaleString();
  return t;
};

const getSentMessages = () => {
  fetch(`${baseUrl}/api/v1/messages/sent`, {
    method: 'GET',
    headers: {
      'x-access-token': token,
    },
  }).then(res => res.json())
    .then((res) => {
      if (res.status === 200) {
        const sentMsg = document.querySelector('#sent-message');
        const { data } = res;
        let num = 0;
        if (data.length) {
          data.forEach((sent) => {
            num += 1;
            sentMsg.innerHTML += `
            <div class="main-flex message-list" onclick="openMessage('sentMail'); getOnesentMessage(${sent.id})">
              <span class="col-3 flex">
                <span class="col-1 arrow-cover flex"><i class="fas fa-arrow-circle-right arrow mr-25"></i>
                  <i class="fas fa-plane-departure dark-col ml-25" title="sent message"></i>
                </span>
                <span class="col-9 mail-head draft-t">To : ${sent.firstname} ${sent.lastname}</span>
              </span>
              <article class="col-7 mail-body">${sent.message} </article>
              <span class="col-2 flex justify-content-sb">
                <span class="col-2 center-text start-text" title="delete"><i class="fas fa-trash delete"></i></span>
                <span class="col-2 center-text start-text retract" title="Retract this Sent Message"><i class="fas fa-undo"></i></span>
                <span class="col-6 center-text start-text">${formatDate(sent.createdon)}</span>
              </span>
            </div>
            `;
          });
        }
      }
      console.log('res', res);
    }).catch((err) => {
      console.log('err =>', err);
    });
};

getSentMessages();
