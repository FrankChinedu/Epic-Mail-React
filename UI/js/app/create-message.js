/* eslint-disable no-unused-vars */
/* eslint-disable no-undef */
const modal = document.createElement('script');
modal.src = './js/app/modal.js';
document.head.appendChild(modal);
const baseUrl = 'http://127.0.0.1:4000';

const token = window.localStorage.getItem('accessToken');

const sendMessage = () => {
  let email = document.querySelector('#receipient').value;
  const subject = document.querySelector('#subject').value;
  const message = document.querySelector('#text-area').value;
  const status = 'sent';
  email = email.trim();

  // copied https://www.w3resource.com/javascript/form/email-validation.php
  const validate = () => {
    const re = /^\w+([\.-]?\w+)*@\w+([\.-]?\w+)*(\.\w{2,3})+$/;
    if (email === '' || !re.test(email)) {
      const str = 'receipient must not be empty and you must be a valid email';
      openModal(str);
      setTimeout(() => {
        closeModal();
      }, 4000);
      return false;
    } if (subject === '') {
      const str = 'subject cannot be empty';
      openModal(str);
      setTimeout(() => {
        closeModal();
      }, 4000);
      return false;
    } if (message === '') {
      const str = 'Body of the email cannot be empty';
      openModal(str);
      setTimeout(() => {
        closeModal();
      }, 4000);
      return false;
    }
    return true;
  };
  const valid = validate(email);

  if (valid) {
    const data = {
      email,
      subject,
      status,
      message,
    };
    fetch(`${baseUrl}/api/v1/messages`, {
      method: 'POST',
      body: JSON.stringify(data),
      headers: {
        'Content-Type': 'application/json',
        'x-access-token': token,
      },
    }).then(res => res.json())
      .then((res) => {
        console.log(res);
        // if (res.status === 200)
        // window.location.href = './inbox-page.html';
      });
  }
};

const saveDraft = () => {
  // console.log('close');
  closeModal();
};
