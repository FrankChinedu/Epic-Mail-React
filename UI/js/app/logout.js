/* eslint-disable no-unused-vars */
/* eslint-disable no-restricted-globals */
/* eslint-disable no-use-before-define */
/* eslint-disable no-undef */
/* eslint-disable no-param-reassign */
