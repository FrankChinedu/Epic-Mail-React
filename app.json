{
	"name": "Epic mail",
	"description": "app for sending and receiving emails",
	"env": {
		"DATABASE_URL": {
			"required": true
		},
		"JWT_SECRET": {
			"required": true
		},
		"NODE_ENV": "review"
	},
	"addons": [
		{
			"plan": "heroku-postgresql",
			"as": "DATABASE_URL",
			"options": {
				"version": "9.5"
			}
		}
	]
}
