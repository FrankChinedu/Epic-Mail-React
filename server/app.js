/* eslint-disable import/no-dynamic-require */
import 'dotenv/config';
import '@babel/polyfill';
import express from 'express';
import bodyParser from 'body-parser';
import cors from 'cors';
import morgan from 'morgan';
import swaggerUi from 'swagger-ui-express';
import YAML from 'yamljs';
import path from 'path';
import fs from 'fs';
import { createAllTables } from './v2/model/index';

const swaggerDocument = YAML.load(`${__dirname}/../swagger.yaml`);

const app = express();
app.use('/api-docs', swaggerUi.serve, swaggerUi.setup(swaggerDocument));

app.use(
  cors({
    origin: '*',
    methods: 'GET, HEAD, PUT, PATCH, POST, DELETE, OPTIONS',
    preflightContinue: false,
    optionsSuccessStatus: 204,
  }),
);

app.all('/*', (req, res, next) => {
  res.header('Access-Control-Allow-Origin', '*');
  res.header('Access-Control-Allow-Headers', 'X-Requested-With');
  next();
});

app.use(bodyParser.urlencoded({ extended: true }));
app.use(morgan('combined'));
app.use(bodyParser.json());
app.use(cors());

app.use('/', (req, res, next) => {
  if (req.originalUrl !== '/') {
    next();
    return;
  }
  res.send({
    message: 'welcome to EPIC MAIL',
  });
});

app.use((req, res, next) => {
  let version = req.url.match(/\/api\/(v[0-9]+).*/) || [];
  version = version[1] || null;

  if (version) {
    const routePath = path.join(__dirname, `${version}/app.js`);

    if (!fs.existsSync(routePath)) {
      return res.status(404).send({
        status: 404,
        error: ['route not found'],
      });
    }
    require(routePath).default(app);
  }
  next();
  return {};
});

const create = () => {
  createAllTables();
};

create();

app.listen(process.env.PORT, () => {
  console.log(`server start at port ${process.env.PORT} `);
});

module.exports = app;
