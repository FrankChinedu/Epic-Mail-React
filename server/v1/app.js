import routes from './routes/api';

const apiURL = '/api/v1';
global.apiURL = apiURL;

export default (app) => {
  Object.keys(routes).forEach((key) => {
    const value = routes[key];
    app.use(`${apiURL}/`, value);
  });
};
