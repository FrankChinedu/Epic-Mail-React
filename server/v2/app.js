import routes from './routes/api';

const apiURL = '/api/v2';
global.apiURL = apiURL;

export default (app) => {
  Object.keys(routes).forEach((key) => {
    const value = routes[key];
    app.use(`${apiURL}/`, value);
  });
};
