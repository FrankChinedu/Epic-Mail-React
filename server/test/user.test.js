import supertest from 'supertest';
import app from '../app';

const server = () => supertest(app);

const { apiURL } = global;

describe(' Sign up', () => {
  describe('sign up', () => {
    it('should create an account for a new user ', async () => {
      const data = {
        firstname: 'frank',
        lastname: 'angelo',
        email: 'fangelo@me.com',
        password: '12345678'
      };
      const { status, body } = await server()
        .post(`${apiURL}/auth/signup`)
        .send(data);
      expect(Object.keys(body.data)).toMatchSnapshot();
      expect(status).toBe(201);
    });
  });

  describe('sign up', () => {
    it('should create an account for a new user ', async () => {
      const data = {
        firstname: 'john',
        lastname: 'doe',
        email: 'johlongn@doe.com',
        password: '12345678'
      };

      const { status, body } = await server()
        .post(`${apiURL}/auth/signup`)
        .send(data);
      expect(Object.keys(body.data)).toMatchSnapshot();
      expect(status).toBe(201);
    });
  });

  describe('/Post auth/signup', () => {
    it('should not be able to sign up a new user if password parameter is missing ', async () => {
      const data = {
        firstname: 'frank',
        lastname: 'angelo',
        email: 'angelo@me.com',
        password: ''
      };
      const { status, body } = await server()
        .post(`${apiURL}/auth/signup`)
        .send(data);

      expect(Object.keys(body)).toMatchSnapshot();
      expect(status).toBe(400);
    });
  });

  describe('/Post auth/signup', () => {
    it('should not be able to sign up a new user if first name parameter is missing', async () => {
      const data = {
        firstname: '',
        lastname: 'angelo',
        email: 'angelo@me.com',
        password: '12345678'
      };
      const { status, body } = await server()
        .post(`${apiURL}/auth/signup`)
        .send(data);
      expect(Object.keys(body)).toMatchSnapshot();
      expect(status).toBe(400);
    });
  });

  describe('/Post auth/signup', () => {
    it('should not be able to sign up a new user if no parameter is missing', async () => {
      const data = {};
      const { status, body } = await server()
        .post(`${apiURL}/auth/signup`)
        .send(data);
      expect(Object.keys(body)).toMatchSnapshot();
      expect(status).toBe(400);
    });
  });

  describe('/Post auth/login', () => {
    it('should log a user in', async () => {
      const data = {
        email: 'fangelo@me.com',
        password: '12345678'
      };
      const { status, body } = await server()
        .post(`${apiURL}/auth/login`)
        .send(data);

      expect(Object.keys(body.data)).toMatchSnapshot();
      expect(status).toBe(200);
    });

    it('should not be able to log a user in if wrong parameters are passed', async () => {
      const data = {
        email: 'franki@me.com',
        password: '123456789'
      };
      const { status, body } = await server()
        .post(`${apiURL}/auth/login`)
        .send(data);
      expect(Object.keys(body)).toMatchSnapshot();
      expect(status).toBe(400);
    });
  });
});
