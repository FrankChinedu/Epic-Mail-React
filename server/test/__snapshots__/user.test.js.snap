// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` Sign up /Post auth/login should log a user in 1`] = `
Array [
  "firstname",
  "lastname",
  "email",
  "token",
]
`;

exports[` Sign up /Post auth/login should not be able to log a user in if wrong parameters are passed 1`] = `
Array [
  "status",
  "error",
]
`;

exports[` Sign up /Post auth/signup should not be able to sign up a new user if first name parameter is missing 1`] = `
Array [
  "status",
  "error",
]
`;

exports[` Sign up /Post auth/signup should not be able to sign up a new user if no parameter is missing 1`] = `
Array [
  "status",
  "error",
]
`;

exports[` Sign up /Post auth/signup should not be able to sign up a new user if password parameter is missing  1`] = `
Array [
  "status",
  "error",
]
`;

exports[` Sign up sign up should create an account for a new user  1`] = `
Array [
  "firstname",
  "lastname",
  "email",
  "token",
]
`;

exports[` Sign up sign up should create an account for a new user  2`] = `
Array [
  "firstname",
  "lastname",
  "email",
  "token",
]
`;
